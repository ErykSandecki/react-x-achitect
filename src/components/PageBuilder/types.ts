// types
import { TRectCoordinates } from 'types';

export type TRectArea = TRectCoordinates | null;
