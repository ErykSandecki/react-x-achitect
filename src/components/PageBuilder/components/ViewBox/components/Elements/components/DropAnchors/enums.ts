export enum DropAnchorsPosition {
  top = 'top',
  bottom = 'bottom',
  left = 'left',
  right = 'right',
}
