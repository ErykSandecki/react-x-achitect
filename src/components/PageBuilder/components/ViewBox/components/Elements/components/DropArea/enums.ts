export enum DropAreaPosition {
  top = 'top',
  bottom = 'bottom',
  left = 'left',
  right = 'right',
}
