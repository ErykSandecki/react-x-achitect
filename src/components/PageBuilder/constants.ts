export const translationNameSpace = 'pageBuilder';

export const TOOLBAR_HEIGHT = 56;
