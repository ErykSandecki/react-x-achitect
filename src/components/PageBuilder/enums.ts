export enum Constrain {
  bottom = 'bottom',
  centerHorizontal = 'centerHorizontal',
  centerVertical = 'centerVertical',
  left = 'left',
  right = 'right',
  top = 'top',
}
