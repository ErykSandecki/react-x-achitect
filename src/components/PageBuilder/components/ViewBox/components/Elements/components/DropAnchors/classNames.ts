import { mapValues } from 'lodash';

// types
import { DropAnchorsPosition } from './enums';

export const className = 'DropAnchors';

export const classNames = {
  [className]: className,
  prompt: {
    name: `${className}__prompt`,
    modificators: {
      ...mapValues(
        DropAnchorsPosition,
        (position) => `${className}__prompt--${position}`,
      ),
    },
  },
  anchor: {
    name: `${className}__anchor`,
    modificators: {
      ...mapValues(
        DropAnchorsPosition,
        (position) => `${className}__anchor--${position}`,
      ),
    },
  },
};
