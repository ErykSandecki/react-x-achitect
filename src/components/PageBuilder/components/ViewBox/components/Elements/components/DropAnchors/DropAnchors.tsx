import { FC, ReactNode } from 'react';
import { useSelector } from 'react-redux';

// components
import { E2EDataAttribute } from 'shared';

// hooks
import { useDropAnchorsEvents } from './hooks/useDropAnchorsEvents';
import { useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';

// store
import {
  eventSelectorCreator,
  isDraggableSelectorCreator,
} from 'store/pageBuilder/selectors';

// styles
import styles from './drop-anchors.scss';

// types
import { DropAnchorsPosition } from './enums';
import { E2EAttribute, TElement } from 'types';
import { MouseMode } from 'components/PageBuilder/enums';

// utils
import { enumToArray } from 'utils';

export type TDropAnchorsProps = {
  children: ReactNode;
  id: TElement['id'];
  index: number;
  mouseMode: MouseMode;
  parentId: TElement['parentId'];
};

const DropAnchors: FC<TDropAnchorsProps> = ({
  children,
  id,
  index,
  mouseMode,
  parentId,
}) => {
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);
  const { onMouseEnter, ...events } = useDropAnchorsEvents(index, mouseMode);
  const isDraggable = useSelector(isDraggableSelectorCreator(id));
  const possibleIndexPosition = useSelector(
    eventSelectorCreator('possibleIndexPosition'),
  ) as number;
  const possibleParent = useSelector(eventSelectorCreator('possibleParent'));
  const isFirst = index === 0;
  const isParent = possibleParent === parentId;
  const displayPrevPrompt =
    !isDraggable && isParent && isFirst && possibleIndexPosition === index;
  const displayNextPrompt =
    !isDraggable && isParent && possibleIndexPosition - 1 === index;

  return (
    <div className={cx(classNamesWithTheme[className])}>
      {displayPrevPrompt && (
        <div
          className={cx(
            classNamesWithTheme.prompt.name,
            classNamesWithTheme.prompt.modificators.top,
          )}
        />
      )}
      {children}
      {displayNextPrompt && (
        <div
          className={cx(
            classNamesWithTheme.prompt.name,
            classNamesWithTheme.prompt.modificators.bottom,
          )}
        />
      )}
      {enumToArray(DropAnchorsPosition).map((position) => (
        <E2EDataAttribute
          key={position as keyof typeof DropAnchorsPosition}
          type={E2EAttribute.anchor}
          value={position as string}
        >
          <div
            className={cx(
              classNamesWithTheme.anchor.name,
              classNamesWithTheme.anchor.modificators[
                position as keyof typeof DropAnchorsPosition
              ],
            )}
            onMouseEnter={() =>
              onMouseEnter(
                DropAnchorsPosition[
                  position as keyof typeof DropAnchorsPosition
                ],
              )
            }
            {...events}
          />
        </E2EDataAttribute>
      ))}
    </div>
  );
};

export default DropAnchors;
