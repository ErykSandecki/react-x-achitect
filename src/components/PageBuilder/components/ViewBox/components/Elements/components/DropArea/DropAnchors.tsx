import { FC, ReactNode } from 'react';
import { useSelector } from 'react-redux';

// hooks
import { useDropAreaEvents } from './hooks/useDropAreaEvents';
import { useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';

// store
import {
  eventSelectorCreator,
  isDraggableSelectorCreator,
} from 'store/pageBuilder/selectors';

// styles
import styles from './drop-anchors.scss';

// types
import { DropAreaPosition } from './enums';
import { MouseMode } from 'components/PageBuilder/enums';
import { TElement } from 'types';

// utils
import { enumToArray } from 'utils';

export type TDropAnchorsProps = {
  children: ReactNode;
  id: TElement['id'];
  index: number;
  mouseMode: MouseMode;
  parentId: TElement['parentId'];
};

const DropAnchors: FC<TDropAnchorsProps> = ({
  children,
  id,
  index,
  mouseMode,
  parentId,
}) => {
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);
  const { onMouseEnter, ...events } = useDropAreaEvents(index, mouseMode);
  const isDraggable = useSelector(isDraggableSelectorCreator(id));
  const possibleIndexPosition = useSelector(
    eventSelectorCreator('possibleIndexPosition'),
  ) as number;
  const possibleParent = useSelector(eventSelectorCreator('possibleParent'));
  const isFirst = index === 0;
  const isParent = possibleParent === parentId;
  const displayPrevPrompt =
    !isDraggable && isParent && isFirst && possibleIndexPosition === index;
  const displayNextPrompt =
    !isDraggable && isParent && possibleIndexPosition - 1 === index;

  return (
    <div className={cx(classNamesWithTheme[className])}>
      {displayPrevPrompt && (
        <div
          className={cx(
            classNamesWithTheme.prompt.name,
            classNamesWithTheme.prompt.modificators.top,
          )}
        />
      )}
      {children}
      {displayNextPrompt && (
        <div
          className={cx(
            classNamesWithTheme.prompt.name,
            classNamesWithTheme.prompt.modificators.bottom,
          )}
        />
      )}
      {enumToArray(DropAreaPosition).map((position) => (
        <div
          className={cx(
            classNamesWithTheme.anchor.name,
            classNamesWithTheme.anchor.modificators[
              position as keyof typeof DropAreaPosition
            ],
          )}
          key={position as keyof typeof DropAreaPosition}
          onMouseEnter={() =>
            onMouseEnter(
              DropAreaPosition[position as keyof typeof DropAreaPosition],
            )
          }
          {...events}
        />
      ))}
    </div>
  );
};

export default DropAnchors;
