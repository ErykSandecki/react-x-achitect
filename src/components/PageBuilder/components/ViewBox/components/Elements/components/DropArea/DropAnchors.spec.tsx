import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

// components
import DropAnchors from './DropAnchors';

// mocks
import {
  elementDynamicDataMock,
  elementStaticDataMock,
  pageBuilderStateMock,
} from 'test/mocks/reducer/pageBuilderMock';
import { wholeStateMock } from 'test/mocks/reducer/wholeStateMock';
import 'test/mocks/sagas/allSagas';

// others
import { REDUCER_KEY as PAGE_BUILDER } from 'store/pageBuilder/actionsType';

// store
import { configureStore } from 'store/store';

// types
import { MouseMode } from 'components/PageBuilder/enums';

const stateMock = {
  ...wholeStateMock,
  [PAGE_BUILDER]: {
    ...pageBuilderStateMock[PAGE_BUILDER],
    elements: {
      dynamicData: { [elementDynamicDataMock.id]: elementDynamicDataMock },
      staticData: { [elementStaticDataMock.id]: elementStaticDataMock },
    },
  },
};

describe('DraggableArea snapshots', () => {
  it('should render DraggableArea', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <DropAnchors
          id={elementDynamicDataMock.id}
          index={0}
          mouseMode={MouseMode.default}
          parentId={elementStaticDataMock.parentId}
        >
          element
        </DropAnchors>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
