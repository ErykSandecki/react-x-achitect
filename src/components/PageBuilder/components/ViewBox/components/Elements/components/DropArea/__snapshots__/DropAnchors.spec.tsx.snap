// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggableArea snapshots should render DraggableArea 1`] = `
<DocumentFragment>
  <div
    class="DropAnchors DropAnchors--light"
  >
    element
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--top DropAnchors__anchor--top--light"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--bottom DropAnchors__anchor--bottom--light"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--left DropAnchors__anchor--left--light"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--right DropAnchors__anchor--right--light"
    />
  </div>
</DocumentFragment>
`;
